remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java

```diff
@@ -236,15 +292,68 @@ private NodeRef changeCategoryName(final NodeRef categoryNodeRef, final String n
     }
 
     /**
-     * Verify if fixed category name is not empty.
+     * Validate if fixed category name is not empty.
      *
      * @param fixedCategoryModel Fixed category model.
      */
-    private void verifyCategoryFields(final Category fixedCategoryModel)
+    private void validateCategoryFields(final Category fixedCategoryModel)
     {
         if (StringUtils.isEmpty(fixedCategoryModel.getName()))
         {
             throw new InvalidArgumentException(NOT_NULL_OR_EMPTY);
         }
     }
+
+    private boolean isRootCategoryPresent(final Collection<NodeRef> categoryNodeRefs)
+    {
+        return categoryNodeRefs.stream().anyMatch(this::isRootCategory);
+    }
+
+    private boolean isCategoryAspectMissing(final NodeRef nodeRef)
+    {
+        return !nodeService.hasAspect(nodeRef, ContentModel.ASPECT_GEN_CLASSIFIABLE);
+    }
+
+    /**
+     * Merge already present and new categories ignoring repeating ones.
+     *
+     * @param currentCategories Already present categories.
+     * @param newCategories Categories which should be added.
+     * @return Merged categories.
+     */
+    private Collection<NodeRef> mergeCategories(final Serializable currentCategories, final Collection<NodeRef> newCategories)
+    {
+        if (currentCategories == null)
+        {
+            return newCategories;
+        }
+
+        final Collection<NodeRef> allCategories = DefaultTypeConverter.INSTANCE.getCollection(NodeRef.class, currentCategories);
+        newCategories.stream()
+            .filter(category -> !allCategories.contains(category))
+            .forEach(allCategories::add);
```

Comment:
Do we know that `allCategories` is modifiable? Instead we could use the union of two sets to determine what the new value should contain.

Reply:
The `.getCollection()` always creates a new `ArrayList<T>` so yes - it's modifiable. In the other hand it's nice suggestion to use sets union - I will change my solution.